lib/spack/spack/util/naming.py

```diff
@@ -6,80 +6,115 @@
+        "for",
+        "from",
+        "global",
+        "if",
+        "import",
+        "in",
+        "is",
+        "lambda",
+        "nonlocal",
+        "not",
+        "or",
+        "pass",
+        "raise",
+        "return",
+        "try",
+        "while",
+        "with",
+        "yield",
+    ]
+)
+
+RESERVED_NAMES_LIST_MIXED_CASE = ("False", "None", "True")
 
 # Valid module names can contain '-' but can't start with it.
-_valid_fully_qualified_module_re = r"^(\w[\w-]*)(\.\w[\w-]*)*$"
+_VALID_MODULE_RE_V1 = re.compile(r"^\w[\w-]*$")
+
+_VALID_MODULE_RE_V2 = re.compile(r"^[a-z_][a-z0-9_]*$")
 
 
-def mod_to_class(mod_name):
-    """Convert a name from module style to class name style.  Spack mostly
-    follows `PEP-8 <http://legacy.python.org/dev/peps/pep-0008/>`_:
+def pkg_name_to_class_name(pkg_name: str):
+    """Convert a Spack package name to a class name, based on
+    `PEP-8 <http://legacy.python.org/dev/peps/pep-0008/>`_:
 
        * Module and package names use lowercase_with_underscores.
        * Class names use the CapWords convention.
 
-    Regular source code follows these convetions.  Spack is a bit
-    more liberal with its Package names and Compiler names:
+    Not all package names are valid Python identifiers:
 
-       * They can contain '-' as well as '_', but cannot start with '-'.
+       * They can contain '-', but cannot start with '-'.
        * They can start with numbers, e.g. "3proxy".
 
-    This function converts from the module convention to the class
-    convention by removing _ and - and converting surrounding
-    lowercase text to CapWords.  If mod_name starts with a number,
-    the class name returned will be prepended with '_' to make a
-    valid Python identifier.
+    This function converts from the package name to the class convention by removing _ and - and
+    converting surrounding lowercase text to CapWords.  If package name starts with a number, the
+    class name returned will be prepended with '_' to make a valid Python identifier.
     """
-    validate_module_name(mod_name)
-
-    class_name = re.sub(r"[-_]+", "-", mod_name)
+    class_name = re.sub(r"[-_]+", "-", pkg_name)
     class_name = string.capwords(class_name, "-")
     class_name = class_name.replace("-", "")
 
-    # If a class starts with a number, prefix it with Number_ to make it
-    # a valid Python class name.
-    if re.match(r"^[0-9]", class_name):
-        class_name = "_%s" % class_name
+    # Ensure that the class name is a valid Python identifier
+    if re.match(r"^[0-9]", class_name) or class_name in RESERVED_NAMES_LIST_MIXED_CASE:
+        class_name = f"_{class_name}"
 
     return class_name
 
 
-def spack_module_to_python_module(mod_name):
-    """Given a Spack module name, returns the name by which it can be
-    imported in Python.
-    """
-    if re.match(r"[0-9]", mod_name):
-        mod_name = "num" + mod_name
+def mod_to_pkg_name(dirname: str, package_api: Tuple[int, int]) -> str:
```

Comment:
Can we call this:
```suggestion
def mod_to_pkg_name(modulename: str, package_api: Tuple[int, int]) -> str:
```
to avoid confusion on directory name vs Python module name?

Reply:
Also, I think we might have some regressions for package API v1.0:
```
>>> from spack.util import naming
>>> naming.mod_to_pkg_name("3dtk", (1, 0))
'3dtk'
>>> naming.mod_to_pkg_name("num3dtk", (1, 0))
'num3dtk'
>>> naming.pkg_name_to_mod("3dtk", (1,0))
'3dtk'
```
On `develop` we have:
```
>>> naming.spack_module_to_python_module("3dtk")
'num3dtk'
>>> naming.possible_spack_module_names("num3dtk")
['3dtk']
```